mlfinlab/backtest_statistics/statistics.py

```diff
@@ -0,0 +1,233 @@
+    negative_concentration = bets_concentration(returns[returns < 0]) 
+    # Concentration of bets/time period (month by default)
+    time_concentration = bets_concentration(returns.groupby(pd.Grouper(freq='M')).count()) 
+    return (positive_concentration, negative_concentration, time_concentration)
+
+
+def compute_drawdown_and_time_under_water(returns: pd.Series, dollars: bool = False) -> tuple:
+    """
+    Snippet 14.4, page 201, Calculates drawdowns and time under water for pd.Series
+    of returns or dollar performance.
+    
+    Intuitively, a drawdown is the maximum loss suffered by an investment between
+    two consecutive high-watermarks. The time under water is the time
+    elapsed between an high watermark and the moment the PnL (profit and loss) 
+    exceeds the previous maximum PnL.     
+    
+    Return details:
+    - Drawdown series index is time of a high watermark and value of a drawdown after.
+    - Time under water index is time of a high watermark and how much time passed till
+    next high watermark in years. 
+
+    :param returns: (pd.Series) returns from bets
+    :param dollars: (bool) flag if given dollar performance and not returns 
+    :return: (tuple of pd.Series) series of drawdowns and time under water 
+                                if dollars, then in dollars, else as a %
+            
+    """
+    frame = returns.to_frame('pnl')
+    frame['hwm'] = returns.expanding().max() # Adding high watermarks as column
+    # Grouped as min returns by high watermarks
+    high_watermarks = frame.groupby('hwm').min().reset_index()
+    high_watermarks.columns = ['hwm','min']
+    # Time high watermark occured
+    high_watermarks.index = frame['hwm'].drop_duplicates(keep='first').index
+    # Picking ones that had a drawdown after high watermark
+    high_watermarks = high_watermarks[high_watermarks['hwm'] > high_watermarks['min']]
+    if dollars:
+        drawdown = high_watermarks['hwm'] - high_watermarks['min']
+    else:
+        drawdown = 1 - high_watermarks['min'] / high_watermarks['hwm']
+    time_under_water = ((high_watermarks.index[1:] - 
+                         high_watermarks.index[:-1]) / np.timedelta64(1,'Y')).values 
+    time_under_water = pd.Series(time_under_water, index = high_watermarks.index[:-1])
+    return drawdown, time_under_water
+
+
+def sharpe_ratio(returns: pd.Series, cumulative: bool = False, 
+                 entries_per_year:int = 252, risk_free_rate:float = 0) -> float:
+    """
+    Calculates Annualized Sharpe Ratio for pd.Series of returns.
+        
+    :param returns: (pd.Series) returns
+    :param cumulative: (bool) flag if returns are cumulative (no by default)
+    :param entries_per_year: (int) times returns are recorded per year (days by default)
+    :param risk_free_rate: (float) risk-free rate (0 by default)
+    :return: (float) Annualized Sharpe Ratio
+    """  
+    if cumulative:
+        returns = returns.diff() # Converting to simple returns
```

Comment:
Is diff correct? I think it assumes your looking at the log cumulative returns. Since log is additive. This wont work if they used the normal cumulative returns. Is this how it is in the book?

Reply:
I was thinking of log returns but didn't state it. My bad. Will make another flag and reversion from normal cumulative returns method. 

The book doesn't specify code for neither of the Sharpe ratios :(

Reply:
I think the standard is to calculate the sharpe in normal returns and not log returns.

Reply:
why to the power of 0.5? 

```**(1/2)```

Reply:
Usually: 

```
sharpe = returns.mean() / returns.std() * np.sqrt(days_in_year)
```

Oh thats why you take the power of 0.5:

Reply:
I've cahnged it to normal and not log returns, left the conversion from cumulative normal returns to normal returns.